File: 94_Binary Tree Inorder Traversal/94_Binary Tree Inorder Traversal.cpp
#include <iostream>
#include <cstdlib>
#include <vector>
#include <stack>

using namespace std;

/**
* Definition for a binary tree node.
* struct TreeNode {
*     int val;
*     TreeNode *left;
*     TreeNode *right;
*     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
* };
*/

struct TreeNode {
	int val;
	TreeNode *left;
	TreeNode *right;
	TreeNode(int x) : val(x), left(NULL), right(NULL){}
};

class Solution {
public:
	vector<int> inorderTraversal(TreeNode* root) {
		//Ê÷¿Õ£¬·µ»Ø¿Õvector
		if (!root)
			return vector<int>();
		
		vector<int> ret;
		stack<TreeNode *> s;
		
		//±éÀú
		TreeNode *p = root;
		do
		{
			while (p != NULL)
			{
				s.push(p);
				p = p->left;
			}

			if (!s.empty())
			{
				p = s.top();
				s.pop();

				ret.push_back(p->val);

				p = p->right;
			}
		} while (!s.empty() || p != NULL);

		return ret;
	}
};

int main()
{

}

File: 94. Binary Tree Inorder Traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
        Stack<TreeNode> stack = new Stack();
        List<Integer> inorderList = new ArrayList<>();
        
        while(!stack.isEmpty() || root != null) {
            while(root != null) {
                stack.push(root);
                root = root.left;
            }
            
            root = stack.pop();
            
            inorderList.add(root.val);
            
            root = root.right;
        }
        
        return inorderList;
    }
    
    
    public List<Integer> inorderTraversal1(TreeNode root) {
        List<Integer> inorderList = new ArrayList<>();
        
        inorderTraversal(root, inorderList);
        return inorderList;
    }
    
    private void inorderTraversal(TreeNode root, List<Integer> inorderList) {
        if (root == null) {
            return;
        }
        
        inorderTraversal(root.left, inorderList);
        inorderList.add(root.val);
        inorderTraversal(root.right, inorderList);
    }
}
